### PR TITLE
Fix broken licenses by searching through readme file, GitHub repo

### PR DESCRIPTION
**Pull request by SelvarajKaruppusamy**
@andrew-mclachlan For the packages that have a missing license file, below are the possible ways (until step 4) to increase license detection according to my current observations,

Below steps would be executed in the following order and if the license file is found in any step then license info is saved and further steps for that module will be skipped.

- [x] 1.  For some modules, the README file would contain the complete license text. The complete license text can be extracted for such modules.
_The [license-checker](https://www.npmjs.com/package/license-checker) tool will return the contents of the README file as the license information if the LICENSE file is missing._

- [x] 2. Fetch the LICENSE file if it is present in the GitHub repo.
_As some modules will not be the latest in our business application, this works if the latest version of that module that contains LICENSE file in repo_

- [x] 3. If the publisher_name and/or published_year is available in the README file, then the license file can be auto-generated with those details.

- [x] 4. If the type of license is found then the license text can be auto-generated.

**Other features:**
- [ ] 5. (optional) Support for user to transform (change/filter object keys) the output data generated from the [license-appendix](https://github.com/andrew-mclachlan/license-appendix) tool.

- [x] 6. Fix licenses only on user action

- [x] 7. Generate a log file including details of packages that have licenses fixed, not fixed

- [ ] 8. (optional) Users can input a reference file with packages and it's license text for desired packages. license-appendix tool will use this file to fix packages for which licenses couldn't be found/fixed.

I have only implemented the first two solutions and will be completing the rest in a couple of days.

Please feel free to suggest if you have different ideas or improvements. It would be great if we are aligned with the approach.

Thanks, Selva.

### CODE REVIEW

**Comment by andrew-mclachlan** [comment]
Thanks for the update! I ran the update on our software.

There were 15 packages that did not have a license file or repository:

license-appendix-review$ grep ": \"<<" test.json  | grep _file | wc -l
      15
license-appendix-review$ grep ": \"<<" test.json  | grep repository | wc -l
      15

and only two that did not have any license text:

license-appendix-review$ grep ": \"<<" test.json  | grep _text | wc -l
       2

`mamacro@0.0.3` (https://www.npmjs.com/package/mamacro/v/0.0.3) and `tz-offset@0.0.1` (https://github.com/node-cron/tz-offset)

The former is a bit tricky because the version that I am using 0.0.3 does not have a link to https://github.com/xtuc/mamacro which has the LICENSE.

I wonder how best to find that one automatically, if at all since can you guarantee that searching GitHub for the name would reveal the repo.?

**Comment by andrew-mclachlan** [comment]
What would be awesome is to show the number of licenses text/descriptions etc not found., so that you can immediately tell if you should curate the output file.

**Comment by SelvarajKaruppusamy** [comment]
I don't have a better idea than searching for a GitHub repo by the module name but this works for most of the packages which have a LICENSE file in the repo. If you have any better way of doing this, I would love to know.

I had thought of something similar to showing fixed license info. With your above inputs, I think we can do something like the following,

- license-appendix should say which modules don't have licenses and how many of them can be fixed using respective mechanisms.
- e.g. xx licenses can be fetched from Github, xx license can be extracted from Readme file, xx licenses cannot be fixed for xx modules. 
- If the user chooses yes, then licenses will be generated using respective mechanisms and a log file will be generated with the information listing the modules which are fixed (grouped by respective mechanisms)
- If the user chooses no, license-appendix generates only the default output.

What do you think?

**Comment by SelvarajKaruppusamy** [comment]
Searching in Github would work only if the repository field has a valid GitHub url. For mamacro@0.0.3, I am not sure how to find the repo URL automatically. I will let you know if I find a solution.

**Comment by andrew-mclachlan** [comment]
> What do you think?

I think this is a great approach. In my opinion, the key thing is to call out that there is some additional manual checking required ("we did our best, please check the following yourself").

In short, let's not worry about being able to catch everything and the method(s) that you've identified will make this wee tool even more useful!

Thanks!

**Comment by andrew-mclachlan** [comment]
> Searching in Github would work only if the repository field has a valid GitHub url.

Agreed. Let us not worry about the cases such as the one I mentioned with `mamacro`. There will always be one-or-two that we may not be able to find, but that's a whole lot better than 10s of licenses that we're needing to check and add after generating the output appendix file.

**Comment by andrew-mclachlan** [comment]
HI @SelvarajKaruppusamy -

Are you wanting me to wait until you've implemented 3 & 4 before clicking Ship it!?

I do wonder if we put out a count of license(s) not found, then that would be helpful.

I've been tested with your PR and it is working well.

Thanks!

**Comment by SelvarajKaruppusamy** [comment]
Sorry, @andrew-mclachlan  I couldn't spend time on this until today as I had to work on some other priorities.

Items 3,4 are completed and pushed. I will continue to work on this and complete the pending tasks.

I have added a few more tasks, see 6-8

**Comment by SelvarajKaruppusamy** [comment]
@andrew-mclachlan I have pushed the changes for all the above features except those are marked as optional. I think the optional features could cause confusion to some users as it requires some javascript knowledge and understanding of the license-appendix tool. I believe for such use cases small scripts can be created by the users themselves.

Let me know if further changes required.

File: index.js [hunk 1]
```diff
@@ -192,16 +193,46 @@ checker.init({ start: program.path, production: true, customFormat: customFormat
           let licenseFile = license.licenseFile.toLowerCase();
           return licenseFile.indexOf('not found') !== -1 || licenseFile.indexOf('readme') !== -1;
         });
-
-        let {fixedLicenses, fixesLog} = await fixLicenseErrors(licenseErrors, program.auth);
-        includeFixes(licenses, fixedLicenses);
+        console.log(`${licenseErrors.length} packages found with incorrect licenses.`);
+
+        let shouldFixLicenses = licenseErrors.length && await (async () => {
```
Comment (andrew-mclachlan) [inline index.js]: I usually run this script as part of a Jenkins build. I assume that this'll block. Can we add a command line option to disable this?

**Comment by andrew-mclachlan** [comment]
Thanks for the updates! I'll test. I do have one thought about fix licenses and automated build systems.

File: index.js (hunk 1 above)
Comment (SelvarajKaruppusamy) [reply index.js]: Sure, Andrew. I will add those options.

**Comment by andrew-mclachlan** [comment]
Curious.. What does `141 packages found having incorrect licenses.` actually mean.

`Licenses couldn't be fixed for 58 packages`. Does that mean I need to curate them. This is more than the first time I ran the script on the same set-up. And if I do a grep, then it tells me that there are 32 occurances of `grep "<<License" test.json | wc -l `

so not sure what the 141 is about and what I need to do for the 58.

Any thoughts?

File: index.js (hunk 1 above)
Comment (SelvarajKaruppusamy) [reply index.js]: Andrew, I have added this feature. Please have a look.

**Comment by SelvarajKaruppusamy** [comment]
> 141 packages found having incorrect licenses

For 141 packages, the `license-text` field contains the README file information and not the expected license information. This is the behavior of the license-checker module.

> Licenses couldn't be fixed for 58 packages.

Yes. Unfortunately, license-appendix couldn't find a license for those packages and you need to curate license for those packages

> Then, why the grep count is not 58?

As I mentioned above, some packages will have content of the README file hence it wouldn't contain the  "License text not found" text in the `license-text` field. If you search for all 58 packages, then you will understand what is happening.


**Additional note:**
This is where feature #8 would be useful. You can curate once and copy the necessary details and save it as a reference file. When you need to generate licenses for the same folder again, you can reuse this file. As I mentioned, I am not sure whether adding this feature would be irrelevant to the license-appendix tool.  Let me know your thoughts.

**Comment by andrew-mclachlan** [comment]
Thanks for the clarification. I will double check. Looks great though!

File: README.md [hunk 2]
```diff
@@ -23,12 +25,25 @@ Create an appendix from the license-checker output.
...
+    `node index.js --fix --json --quiet --path <src_dir> --output licenses.json`
+
+6. Use Oauth token to fetch license from GitHub
+
+   `node index.js --fix --json --path <src_dir> --output licenses.json --auth <your_github_personal_acces_token>`
 
       [Visit this link](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) to create your personal access token.
+
+## Found a bug?
+Please create an issue with the relevant details. Submitting a PR would be even great !!
```
Comment (andrew-mclachlan) [inline README.md:49]: Love it! 💯

**Comment by andrew-mclachlan** [comment]
Thanks for all your efforts! This is brilliant!

**Comment by SelvarajKaruppusamy** [comment]
You are most welcome, @andrew-mclachlan. Glad to help😃

**Pull request merged**